src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests2.cs

```diff
@@ -2596,5 +2596,197 @@ public void M(string s)
+class B : C { }
+class D
+{
+    public static implicit operator D(C c) => throw null;
+    public static implicit operator D(short s) => throw null;
+}
+class E
+{
+    public static implicit operator C(E c) => throw null;
+}
+class F
+{
+    public static implicit operator C(F c) => throw null;
+}
+";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularWithPatternCombinators).VerifyDiagnostics(
+                // (11,15): error CS8506: No best type was found for the switch expression.
+                //         _ = i switch // 1
+                Diagnostic(ErrorCode.ERR_SwitchExpressionNoBestType, "switch").WithLocation(11, 15),
+                // (25,18): error CS0029: Cannot implicitly convert type 'E' to 'D'
+                //             0 => new E(), // 3.1
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "new E()").WithArguments("E", "D").WithLocation(25, 18),
+                // (26,18): error CS0029: Cannot implicitly convert type 'F' to 'D'
+                //             1 => new F(), // 3.2
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "new F()").WithArguments("F", "D").WithLocation(26, 18),
+                // (63,18): error CS0029: Cannot implicitly convert type 'E' to 'D'
+                //             0 => new E(), // 9.1
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "new E()").WithArguments("E", "D").WithLocation(63, 18),
+                // (64,18): error CS0029: Cannot implicitly convert type 'F' to 'D'
+                //             1 => new F(), // 9.2
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "new F()").WithArguments("F", "D").WithLocation(64, 18)
+                );
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+            var switches = tree.GetRoot().DescendantNodes().OfType<SwitchExpressionSyntax>().ToArray();
+            for (int i = 0; i < switches.Length; i++)
+            {
+                var @switch = switches[i];
+                var typeInfo = model.GetTypeInfo(@switch);
+                var conversion = model.GetConversion(@switch);
+                switch (i)
+                {
+                    case 0:
+                    case 7:
+                        Assert.Null(typeInfo.Type);
+                        Assert.Equal("C", typeInfo.ConvertedType.ToTestDisplayString());
+                        Assert.Equal(ConversionKind.SwitchExpression, conversion.Kind);
+                        break;
+                    case 1:
+                        Assert.Null(typeInfo.Type);
+                        Assert.True(typeInfo.ConvertedType.IsErrorType());
+                        Assert.Equal(ConversionKind.Identity, conversion.Kind);
+                        break;
+                    case 2:
+                    case 8:
+                        Assert.Null(typeInfo.Type);
+                        Assert.Equal("D", typeInfo.ConvertedType.ToTestDisplayString());
```

Comment:
>Assert.Equal("D", typeInfo.ConvertedType.ToTestDisplayString()); [](start = 24, length = 64)

This looks unexpected for the explicit cast case.

Reply:
The switch expression conversion is modeled as always occurring on the switch expression itself. As a consequence the switch expression conversion occurs as an argument to the cast.

---
In reply to: [440448752](https://github.com/dotnet/roslyn/pull/45069#discussion_r440448752) [](ancestors = 440448752)